_posts/2023-02-10-cyclonedx-maven-plugin-adventure-continues.md

```diff
@@ -0,0 +1,325 @@
+
+```
+<component type="library" bom-ref="pkg:maven/commons-io/commons-io@2.11.0?type=jar">
+```
+
+In my pull request we discover **commons-io:2.11.0** in the set of resolved artifacts, including it as a known component.  When creating the dependency tree we discover the **commons-io:LATEST** dependency and include it in the tree.  At the end of the process we drop components which are not mentioned in the dependency tree, in this instance the **commons-io:2.11.0** component, but keep the dependency relationship between **webwolf** and **commons-io:LATEST**, which is a missing component.  The bom looks as follows
+
+```
+<dependency ref="pkg:maven/org.owasp.webgoat/webwolf@v8.0.0.M15?type=jar">
+  <dependency ref="pkg:maven/commons-io/commons-io@LATEST?type=jar"/>
+</dependency>
+<dependency ref="pkg:maven/commons-io/commons-io@LATEST?type=jar"/>
+```
+
+## Summarising the issues
+We have three different scenarios here, however each has the same root cause.  Maven is returning a dependency graph which includes marker nodes referencing the original artifact versions and not the versions resolved within the context of the build.  These marker nodes have a different identity to the resolved dependencies and are, therefore, treated separately.  As we would expect, identity does matter!
+
+With the upstream codebase we see the resolved components being included in the bom, but with certain dependency relationships missing from the dependency tree.
+
+With my pull request we see some missing components from the bom, but with all dependency relationships included in the dependency tree.  The problem is that some of these relationships reference dependencies with their original version and not the version resolved by maven.
+
+---
+
+**Note:** The issues we are seeing do not happen with dependencies which have their version managed, if we look at the node for a managed dependency we can see the version of the marker has been updated
+
+```
+org.slf4j:slf4j-api:1.7.25:compile    (org.slf4j:slf4j-api:jar:1.7.25:compile - version managed from 1.6.6; omitted for duplicate)
+```
+
+In this case the version of the marker node has been updated from **1.6.6** to **1.7.25**.
+
+Unfortunately this additional information is not available to us other than through the **toNodeString** method on the **VerboseDependencyNode** class, that is unless we delve under the covers and work on the internal _aether_ dependency tree which does contain a data map including this information.
```

Comment:
Is _aether_ referring to [Aether](https://wiki.eclipse.org/Aether/What_Is_Aether)?  
If so should we provide a link perhaps to clarify this?

Reply:
> Is _aether_ referring to [Aether](https://wiki.eclipse.org/Aether/What_Is_Aether)? If so should we provide a link perhaps to clarify this?

It does.  I included a link in the first post to aether and can do so again.  I'll push an update.

Reply:
This should be updated now

Reply:
Great, thanks!